js/app/modules/selection/articleContext.js

```diff
@@ -13,37 +13,47 @@ const ArticleContext = new Module.Class({
 
     _init: function (props={}) {
         this.parent(props);
-        this._item = {};
-    },
-
-    // Selection.Xapian implementation
-    on_history_changed: function (history) {
-        let item = history.get_current_item();
-        if (!item)
-            return;
-
-        if ((item.query && item.query !== this._item.query) ||
-            (item.page_type === Pages.SET && this._item.model != item.model)) {
-            this._item = item;
+        this._use_query = false;
+        this._use_set = false;
+
+        let history = HistoryStore.get_default();
+        history.connect('notify::current-query', () => {
+            if (history.current_query) {
+                this._use_query = true;
+                this._use_set = false;
+            } else {
+                this._use_query = false;
+            }
             this._set_needs_refresh(true);
-        }
+        });
+        history.connect('notify::current-set', () => {
```

Comment:
can you not combine both these callbacks into one? That is called both for `notify::current-set` and for `'notify::current-query`?